### ISSUE TITLE
Cross-link extensions on class where they apply

### ISSUE THREAD

**Issue by munificent** [dart-lang/dartdoc#2021]
Issue #2001 is about documenting extensions themselves at the point they are defined. That's canonical "place" for an extension's docs.

One concern some users have is that users will run into this scenario.

1. They see some code like `foo.bar()`. They want to know what `bar()` is.
2. They know `foo` is a Foo, so they look up the API docs for Foo.
3. No `bar()`.

What they don't realize is that `bar()` is an extension method defined elsewhere. In the general case, this can't be solved. Users can define extensions on types outside of their own packages, even on core library types or non-class types that have no docs.

But in many cases, the type the extension applies to is a class in the same package where the extension itself is defined. In those cases, it would be nice if the dartdoc for that class could cross-link to the relevant extensions. For example, the C# docs for the IEnumerable interface [have a section for extension methods the .NET core library defines on that type](https://docs.microsoft.com/en-us/dotnet/api/system.collections.ienumerable?view=netframework-4.8#extension-methods).

cc @mit-mit

**Comment by jcollins-g** [comment]
Cases like our dart:ffi's Pointer, where extension methods are being used as a primitive sort of method overloading, may require some additional thought as expanding all possible methods on the class page could be unreadable.  A followup to #2053 will have a first pass idea of what that might look like.

**Comment by jakemac53** [comment]
Any chance we can circle back to this? The list of applicable extensions as it exists today is easy to miss, and requires opening N different pages to see the extension methods.

The .net example linked above I think  is much nicer. I am not sure why from a user perspective why we wouldn't treat extension methods in exactly the same way as regular methods? They could live in a different section to be sure, but it is reasonable to assume extensions from the same package (or maybe, library) are meant to be first class citizens.

**Comment by parlough** [comment]
I think this is super important now with `dart:js_interop` set up the way it is.

I just think some UX questions remain. For example, `Pointer` has a **bunch** of extensions like `Int8Pointer` on `Pointer<Int8>` or the equivalent. Each of these has 4-7 or so properties/operators/methods that are often the same name as other extensions.

Cases like these will probably need some form of grouping and collapsing or something.

**Comment by srawlins** [comment]
Yeah let's prioritize this as P1 for the next release.

**Comment by srawlins** [comment]
I'm getting my ducks in a row to land this. There's a technical complication with `allModelElements`. It might be important that a class's `allModelElements` now include potentially applicable extensions, but doing so would mean that, in order to calculate one class's `allModelElements`, you must know about all extensions first. This violates a couple properties of the package graph. So I need to sort things out a bit.

**Comment by bwilkerson** [comment]
Any updates?

**Comment by srawlins** [comment]
Famous last words in February:

> I'm getting my ducks in a row to land this.

I have hit another issue, https://github.com/dart-lang/dartdoc/issues/3748, and I believe that fixing #3748 will unblock a number of features that have been requested, including this one.

**Comment by bwilkerson** [comment]
Any progress?

**Comment by srawlins** [comment]
There has not been direct progress. I have only found more yaks to shave in order to implement this. I think it should still be P1. And I am working on the other yaks. I should try to link back here where applicable, to leave a paper trail.

**Comment by srawlins** [comment]
Still making slow progress 😁  Removed duplicate notions of "canonical library" last week.

**Comment by srawlins** [comment]
Still slow progress. But the work towards doc-imports actually helped move this forward a bit, as some bugs were discovered in how extensions were handled.

**Comment by srawlins** [comment]
Sorry for the delay, all, but I have a CL that's just about ready to go. There are some design questions though, and I'd love any and all input. @kevmoo I'm looking at you. CC @parlough 

We don't have to do a design research project, or get it perfect the first time. I'd love to move forward with something that works well enough, and doesn't use style that directly clashes with anything else on the site or our other properties.

### Class body layout

On a "class page", put extension instance methods (and properties and operators) inline with "directly declare" instance methods? Or have a bunch of new sections, "Extension instance methods," "Extension instance properties," "Extension operators?" I vote the former.

And how can we distinguish that one is _only available via an extension you may need to import?_ Or that one is _only available on some subtypes?_ For example a field that is available on List (so should be listed on the List page), but actually only available on `List<Enum>` or `Iterable<Future>`?  Here are some shots of what that can look like:

<img width="589" alt="Screenshot 2024-08-27 at 1 48 06 PM" src="https://github.com/user-attachments/assets/79c9a908-4e29-4153-8dcb-6ad8c3cfc912">

---

<img width="482" alt="Screenshot 2024-08-27 at 1 48 21 PM" src="https://github.com/user-attachments/assets/0e4b6051-ce1a-409f-a95c-17ba9a73983e">

### Sidebar layout

Should extension methods (and properties and operators) be inline with other instance members, or in separate sections? Again, I think inline is good. So how to distinguish? Here is one possibility, using simple asterisks and hover text:

<img width="252" alt="Screenshot 2024-08-28 at 9 06 29 AM" src="https://github.com/user-attachments/assets/9c57d325-d32b-48a8-9e97-463f6be4fa7d">

**Comment by kevmoo** [comment]
Hrm. Interesting. My only (slight) concern is that this well only show...what...the default extensions available from the library in question?

Fill folks me confused why some show up and others don't (based on their imports)?

**Comment by munificent** [comment]
I also vote for inline and like the "Available on ..., provided by the ... extension." design and links.

In the sidebar, maybe something like "(ext)"?

**Comment by srawlins** [comment]
> Hrm. Interesting. My only (slight) concern is that this well only show...what...the default extensions available from the library in question?

_Actually_, as I've implemented it, it's pretty wide-sweeping. So in the Dart SDK, we would cross-link any extensions declared anywhere in the Dart SDK (any public ones, that is). But in the Flutter SDK, it's wider: we would cross-link any extensions declared anywhere in any of the documented packages. Good? I... think so?

For example, the collection package is documented at api.flutter.dev. This package provides many extensions, like [IterableIntegerExtension](https://api.flutter.dev/flutter/package-collection_collection/IterableIntegerExtension.html). So this extension exposes members to Iterable expressions. So on every page for a subtype of Iterable, like List, Queue, etc., even the [Characters class](https://api.flutter.dev/flutter/characters/Characters-class.html) from the characters package.

This is not out-of-line with other cases of this "universality" at api.flutter.dev, like how [Iterable's "Implementors"](https://api.flutter.dev/flutter/dart-core/Iterable-class.html) include PathMetrics, CachingIterable, and Characters.

> In the sidebar, maybe something like "(ext)"?

@pq suggested something similar, rather than the rather ambiguous asterisk. I will prototype something like a subscript "e" or "(ext)" or "(E)".

Thanks for everyone's input 🙏

**Comment by mit-mit** [comment]
@srawlins is there a link to Flutter API docs that includes this so we can poke around a bit and see the result?

@goderbauer fyi
